sign/rpmsignverity.c

```diff
@@ -0,0 +1,234 @@
+/**
+ * Copyright (C) 2020 Facebook
+ *
+ * Author: Jes Sorensen <jsorensen@fb.com>
+ */
+
+#include "system.h"
+
+#include <rpm/rpmlib.h>		/* RPMSIGTAG & related */
+#include <rpm/rpmlog.h>		/* rpmlog */
+#include <rpm/rpmfi.h>
+#include <rpm/rpmpgp.h>		/* rpmDigestLength */
+#include "lib/header.h"		/* HEADERGET_MINMEM */
+#include "lib/header_internal.h"
+#include "lib/rpmtypes.h"	/* rpmRC */
+#include <libfsverity.h>
+#include <rpmio/rpmio_internal.h>
+#include "lib/rpmvs.h"
+
+#include "sign/rpmsignverity.h"
+
+#define MAX_SIGNATURE_LENGTH 1024
+
+static int rpmVerityRead(void *opaque, void *buf, size_t size)
+{
+	int retval;
+	rpmfi fi = (rpmfi)opaque;
+
+	retval = rpmfiArchiveRead(fi, buf, size);
+
+	if (retval > 0)
+		retval = 0;
+	return retval;
+}
+
+static char *rpmVeritySignFile(rpmfi fi, size_t *sig_size, char *key,
+			       char *keypass, char *cert, uint16_t algo,
+			       uint32_t block_size)
+{
+    struct libfsverity_merkle_tree_params params;
+    struct libfsverity_signature_params sig_params;
+    struct libfsverity_digest *digest = NULL;
+    rpm_loff_t file_size;
+    char *digest_hex, *sig_hex = NULL;
+    uint8_t *sig = NULL;
+    int status;
+
+    if (S_ISLNK(rpmfiFMode(fi)))
+	file_size = 0;
```

Comment:
No signatures for symlinks? Symlink pointing to an unintended place can have pretty drastic consequences...

Reply:
I totally agree, unfortunately the kernel doesn't currently offer symlink signature support. It is something I want to address there.

Reply:
Ack.